.github/workflows/ci.yml

```diff
@@ -0,0 +1,139 @@
+        branches:
+            - "main"
+            - "release/*"
+    release:
+        types: [created]
+
+jobs:
+    lint:
+        name: Lint Python
+        runs-on: ubuntu-latest
+        strategy:
+            matrix:
+                python-version: ["3.10", "3.11", "3.12"]
+        steps:
+            - name: Checkout
+              uses: actions/checkout@v4
+
+            - name: Cache dependencies
+              uses: actions/cache@v4
+              with:
+                path: ~/.cache/pip
+                key: ${{ runner.os }}-pip-${{ hashFiles('**/pyproject.toml') }}
+                restore-keys: |
+                  ${{ runner.os }}-pip-
+
+            - name: Set up Python
+              uses: actions/setup-python@v5
+              with:
+                python-version: ${{ matrix.python-version }}
+
+            - name: Install dependencies
+              run: pip install .[dev]
+
+            - name: Run Linting
+              run: ruff check src/ tests/
+
+    mypy:
+        name: Mypy
+        runs-on: ubuntu-latest
+        strategy:
+            matrix:
+                python-version: ["3.10", "3.11", "3.12"]
+        steps:
+            - name: Checkout
+              uses: actions/checkout@v4
+
+            - name: Cache dependencies
+              uses: actions/cache@v4
+              with:
+                path: ~/.cache/pip
+                key: ${{ runner.os }}-pip-${{ hashFiles('**/pyproject.toml') }}
+                restore-keys: |
+                  ${{ runner.os }}-pip-
+
+            - name: Set up Python
+              uses: actions/setup-python@v5
+              with:
+                python-version: ${{ matrix.python-version }}
+
+            - name: Install dependencies
+              run: pip install .[dev]
+
+            - name: Run Mypy
+              run: mypy src/ tests/
+
+    test:
+        name: Test Python
+        runs-on: ubuntu-latest
+        strategy:
+            matrix:
+                python-version: ["3.10", "3.11", "3.12"]
+        steps:
+            - name: Checkout
+              uses: actions/checkout@v4
+
+            - name: Cache dependencies
+              uses: actions/cache@v4
+              with:
+                path: ~/.cache/pip
+                key: ${{ runner.os }}-pip-${{ hashFiles('**/pyproject.toml') }}
+                restore-keys: |
+                  ${{ runner.os }}-pip-
+
+            - name: Set up Python
+              uses: actions/setup-python@v5
+              with:
+                python-version: ${{ matrix.python-version }}
+
+            - name: Install dependencies
+              run: pip install .[dev]
+
+            - name: Run Tests
+              run: pytest --cov=src --cov-report=xml tests/ -m "not requires_data"
+
+            - name: Upload Coverage Report
+              uses: actions/upload-artifact@v4
+              with:
+                name: coverage-report-${{ matrix.python-version }}
+                path: coverage.xml
+
+    build:
```

Comment:
Do we need trigger `build` every time when push to `main`?

Reply:
No I agree we should probably only trigger build on release. There are two ways we could accomplish this - either 1) add a conditional line to the same file so that `build` job only triggers when the event is release, or 2) we can separate `build` into a new workflow/config file (e.g. `build.yml`). I kinda like the latter because it makes it obvious that we have two sets of jobs which run in different situations, but not opinionated